tests/Rule/IpTest.php

```diff
@@ -527,16 +557,99 @@ public function dataValidationFailed(): array
                 [new Ip(ranges: ['10.0.0.1', '!10.0.0.0/8', '!babe::/8', 'any'])],
                 ['' => [$notInRangeMessage]],
             ],
+            'custom not in range message' => [
+                '10.0.0.2',
+                [
+                    new Ip(
+                        notInRangeMessage: 'Custom not in range message.',
+                        ranges: ['10.0.0.1', '!10.0.0.0/8', '!babe::/8', 'any'],
+                    ),
+                ],
+                ['' => ['Custom not in range message.']],
+            ],
+            'custom not in range message with parameters' => [
+                '10.0.0.2',
+                [
+                    new Ip(
+                        notInRangeMessage: 'Attribute - {attribute}, value - {value}.',
+                        ranges: ['10.0.0.1', '!10.0.0.0/8', '!babe::/8', 'any'],
+                    ),
+                ],
+                ['' => ['Attribute - , value - 10.0.0.2.']],
+            ],
+            'custom not in range message with parameters, attribute set' => [
+                ['data' => '10.0.0.2'],
+                [
+                    'data' => new Ip(
+                        notInRangeMessage: 'Attribute - {attribute}, value - {value}.',
+                        ranges: ['10.0.0.1', '!10.0.0.0/8', '!babe::/8', 'any'],
+                    ),
+                ],
+                ['data' => ['Attribute - data, value - 10.0.0.2.']],
+            ],
 
-            ['192.168.005.001', [new Ip()], ['' => [$message]]], // Leading zeroes are not supported
+            'leading zeroes' => ['192.168.005.001', [new Ip()], ['' => [$message]]],
```

Comment:
Why don't support leading zeroes?
Suggest create issue and add support.

Reply:
> Why don't support leading zeroes?

I don't know, but it's out of scope of this PR.

> Suggest create issue and add support.

Created the issue - #387.